### ISSUE TITLE
[question] CMakeDeps does not seem to be propagating link dependencies

### ISSUE THREAD

**Issue by jasonbeach** [conan-io/conan#14970]
### What is your question?

Hi! I'm not sure if this is a bug or I'm doing something wrong. I have a small "common" library that depends on `spdlog` and a generic imu library that depends on my common library and specific vectornav library that depends on the generic imu library. so spdlog -> commonlib -> imu_lib -> vectornav_lib.

In the conanfile.py in my common lib, I pull in spdlog with `transitive_headers=True` so it propagates to consumers as it should. similarly, imu_lib pulls in the common lib and the vectornav_hw lib pulls in the imu_hw lib all with `transitive_headers=True`. Compiling succeeds, however the test app for the vectornav lib fails with a linker error 

```
/usr/bin/ld: src/CMakeFiles/vectornav_plugin_test.dir/vectornav_plugin.test.cpp.o: undefined reference to symbol '_ZNK6spdlog7details10backtracer7enabledEv'
```
Sure enough, if I look at the line for building the test app, spdlog is not being linked to:

```
/usr/bin/c++ -m64 -O3 -DNDEBUG -m64 src/CMakeFiles/vectornav_plugin_test.dir/vectornav_plugin.test.cpp.o -o src/vectornav_plugin_test -L/home/ubuntu/.conan2/p/catchc1ea41280289f/p/lib   -L/home/ubuntu/.conan2/p/b/imu76b5606193f31/p/lib   -L/home/ubuntu/.conan2/p/b/contr796d55f401d46/p/lib   -L/home/ubuntu/.conan2/p/proto863068fcaa2b4/p/lib   -L/home/ubuntu/.conan2/p/zeromf037b268a506a/p/lib   -L/home/ubuntu/.conan2/p/b/commonb051c979d4da0/p/lib   -L/home/ubuntu/.conan2/p/b/spdloa496be8bad6b7/p/lib   -L/home/ubuntu/.conan2/p/b/fmt2e17de18e3535/p/lib   -L/home/ubuntu/.conan2/p/b/yaml-f4b34e41c576b/p/lib -Wl,-rpath,/home/ubuntu/.conan2/p/catchc1ea41280289f/p/lib:/home/ubuntu/.conan2/p/b/react76b5606193f31/p/lib:/home/ubuntu/.conan2/p/b/contr796d55f401d46/p/lib:/home/ubuntu/.conan2/p/proto863068fcaa2b4/p/lib:/home/ubuntu/.conan2/p/zeromf037b268a506a/p/lib:/home/ubuntu/.conan2/p/b/raptob051c979d4da0/p/lib:/home/ubuntu/.conan2/p/b/spdloa496be8bad6b7/p/lib:/home/ubuntu/.conan2/p/b/fmt2e17de18e3535/p/lib:/home/ubuntu/.conan2/p/b/yaml-f4b34e41c576b/p/lib  /home/ubuntu/.conan2/p/catchc1ea41280289f/p/lib/libCatch2Main.a  /home/ubuntu/.conan2/p/catchc1ea41280289f/p/lib/libCatch2.a  /home/ubuntu/.conan2/p/b/imu76b5606193f31/p/lib/libimu_hardware.so  -lrt  -lstdc++fs  -lpthread  /home/ubuntu/.conan2/p/b/yaml-f4b34e41c576b/p/lib/libyaml-cpp.a  -lm
``` 
in fact the common utility library is missing as well. (interestingly though both ARE included as link directories just the library themselves don't show up with the other libraries to be linked to.

If I add those two libraries as requirements (literally just add them the `requirements` method of the conanfile.py) of my the vectornav library, but don't modify anything else (e.g. the CMakeLists.txt) then everything builds as it ought to. 

I'll see if I can distill this down to a simpler example.  I've poured over the requirement_traits documentation and as far as I can tell I have everything right with just setting `transitive_headers=True` on the requirement. Is there anything I could be missing?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
i @jasonbeach 

Thanks for your question
> In the conanfile.py in my common lib, I pull in spdlog with transitive_headers=True so it propagates to consumers as it should. similarly,

This  is only necessary if your ``common`` public headers do ``#include`` the ``spdlog`` headers. If only your .cpp implementation files include ``spdlog`` it is not necessary.

I think the most common issue is not specifying the ``package_type`` of your libraries, are you doing it? It can be done in 2 ways:
- Implicitly, declaring an ``option`` that is called ``shared``. Based on the value, the package-type will be deduced
- Explicitly, using ``package_type = "static-library"`` (or shared-library or header-library)

With these declare, it shouldn't be necessary to do ``transitive_libs=True`` in most cases, but if the ``transitive_headers=True`` and those header explicitly require some implementation in the library, and there is some embedding (like a shared library linking a static library), then it might be necessary ``transitive_libs=True``, but those should be the exceptions if APIs are well designed.

Can you please check that? Of course a reproducible use case would be great if that doesn't help.

**Comment by jasonbeach** [comment]
Hi @memsharded,
  It looks like I've been using the implicit method... i.e  for my common library the `conanfile.py` has
```
  options = {"shared": [True, False], "fPIC": [True, False], "GeographicLib": [True, False], "YamlSupport": [True, False]}
  default_options = {"shared": True, "fPIC": True, "GeographicLib": True, "YamlSupport": True}
```
the downstream libraries are similar. 

in my common library I have basically a super thin wrapper  around spdlog that sets up a couple sinks to make out of the box ready to use.  most of the use in downstream libraries is just the spdlog API itself.

I'm trying to come up with a simpler example, but what I've come up with so far actually builds completely, even though if I force ninja to be verbose the invoked compiler commands look like they're similar (i.e. missing spdlog).  So I guess with what I've come up with the linkers able to resolve all the symbols even though the spdlog library wasn't explicitly given.   

This isn't unique to spdlog. I have a function in my common library that demangles a type and I'm getting a similar linker error for it. It looks like it's getting pulled in by a function template (also in my common library) that retrieves a yaml value, so the function template (and hence definition) is in the header, but the demangle function  is defined normally (definition in .cpp file). And I do use that yaml function in my final test app that's failing. Hmmm.

**Comment by jasonbeach** [comment]
So thinking about this a little more and what I assume @memsharded  you're referring to when you say if the "API is well designed" means that a dependency doesn't bleed through, i.e. if library B depends on library A, then ideally library A would only be included in library B's .cpp files so that when library C `#includes` library B's headers, library A's headers don't bleed through into library C. 

However sometimes that's exactly what you want and `spdlog` is a great example of that...it depends on `libfmt` for formatting and `libfmt`'s API is integral to `spdlog`'s API. Looking at `spdlog`'s conan recipe, when they pull in fmt, they set `transitive_headers` AND `transitive_libs` to True I would assume to reflect this melding of APIs so that users of `spdlog` only have bring in `spdlog` to get the whole unified package. 

I set `transitive_libs=True` for `spdlog` in my common library because the goal with it's logger API was to just lightly wrap it (i.e. make it easier to create a logger with a console and file sink), but my goal is that spdlog's API is melded in with my  common library's API so users don't have to really think about where the logger is coming from.  That seems to have fixed the link error for spdlog. For the other linker error with the demangler function, my derived class is directly calling a function from the common library so its correct for it to require the common library directly (and not just be lazy and let the common library get brought in from the base imu package). 

It would be nice if the documentation [here](https://docs.conan.io/2.0/reference/conanfile/methods/requirements.html#package-type-trait-inferring) was more explicit for the default settings for traits for the different package types and requires (I thought that transitive_libs=True was the default and I thought I had traced my way through the conan Requirements class to verify that but apparently I was wrong or missed something.)
